src/librustc_expand/mbe/macro_parser.rs

```diff
@@ -877,7 +877,7 @@ fn parse_nt_inner<'a>(p: &mut Parser<'a>, sp: Span, name: Symbol) -> PResult<'a,
             None => return Err(p.struct_span_err(p.token.span, "expected a statement")),
         },
         sym::pat => token::NtPat(p.parse_pat(None)?),
-        sym::expr => token::NtExpr(p.parse_expr()?),
+        sym::expr => token::NtExpr(p.do_parse_expr()?),
```

Comment:
I don't think the force capture flag and `parse_expr`/`do_parse_expr` separation is necessary, these two calls to `parse_item` and `parse_expr` in `macro_parser.rs` just need to be wrapped into `collect_tokens`.

Reply:
There's one annoying detail though - the collected tokens must not include outer attributes.

Two possible cases are possible here:
- The parsed expression doesn't have outer attributes. This is good, collect the tokens collected here to it.
- The parsed expression has outer attributes. This is also good, it should already have the tokens collected elsewhere.

Actually it should be even simpler - if the parsed expression already has assigned tokens, keep them, otherwise give it tokens collected here in `macro_parser.rs`.